content/existential-containers.md

```diff
@@ -0,0 +1,288 @@
+Similar to Swift, existential containers in Rust are considered supertypes of the types conforming to their bound.
+
+Existential contains are also featured in Haskell, under the [`ExistentialQuantification`](https://wiki.haskell.org/Heterogenous_collections) extension.
+Unlike in Swift and Rust, packing and unpacking in and out of existential containers requires more boilerplate:
+
+```haskell
+{-# LANGUAGE ExistentialQuantification #-}
+
+class Polygon a where
+  area :: a -> Double
+
+data AnyPolygon = forall a . Polygon a => MakePolygon a
+pack :: Polygon a => a -> AnyPolygon
+pack = MakePolygon
+
+instance Polygon Square where
+  area s = 1.0
+instance Polygon Hexagon where
+  area h = 1.0
+
+largest :: [AnyPolygon] -> Maybe AnyPolygon
+largest (x : xs) = case (largest xs) of
+    Nothing -> Just x
+    Just(y) -> Just (if (f x) < (f y) then y else x)
+  where f (MakePolygon a) = area a
+largest [] = Nothing
+```
+
+A more formal exploration of the state of the art as been documented in a research paper presented prior to this SIP [2].
+
+## Alternatives considered
+
+### No change to the language
+
+As already mentioned, Scala's type system is currently strong enough to support the definition of existential containers.
+Hence, no change to the language is strictly necessary to support them.
+The gain offered by the proposal is that method selection on an existential container will look more familiar.
+
+### Converting to and from wrappers
+
+As mentioned in our motivations, the problem of creating a heterogeneous collections can be addressed by defining custom wrappers.
+One can further define implicit conversions to alleviate the syntactic burden.
+This approach is nonetheless strictly more verbose since it requires the definition of a specific container for each type class used in conjunction with heterogeneous collections.
+In that sense, existential containers can be understood as a generalized wrapper.
+
+## FAQ
+
+#### Is there any significant performance overhead in using existential containers?
+
+On micro benchmarks testing method dispatch specifcally, we have measured that dispatching through existential containers in Scala was about twice as slow as traditional virtual method dispatch, which is explained by the extra pointer indirection introduced by an existential container.
+This overhead drops below 10% on larger, more realistic benchmarks [2].
+
+## References
+
+1. Stefan Wehr and Peter Thiemann. 2011. JavaGI: The Interaction of Type Classes with Interfaces and Inheritance. ACM Transactions on Programming Languages and Systems 33, 4 (2011), 12:1–12:83. https://doi.org/10.1145/1985342.1985343
+2. Dimi Racordon and Eugene Flesselle and Matt Bovel. 2024. Existential Containers in Scala. ACM SIGPLAN International Conference on Managed Programming Languages and Runtimes, pp. 55-64. https://doi.org/10.1145/3679007.3685056
+
+
+## Appendix
+
+The following is a possible implementation of existential containers.
```

Comment:
Here's a version that doesn't rely on the experimental modularity language import (and doesn't require the typeclass to extend `TypeClass`):
```scala
/** A value together with an evidence of its type conforming to some type class. */
sealed trait Containing[TC[_]]:
  /** The type of the contained value. */
  type Value: TC as witness

  /** The contained value. */
  val value: Value

object Containing:

  /** Wraps a value of type `V` and a corresponding typeclass instance `TC[V]` into a `Containing[TC]`. */
  def apply[TC[_]](v: Any)[V >: v.type](using TC[V]): Containing[TC] =
    new Containing[TC]:
      type Value >: V <: V
      val value: Value = v
```